plugins/Compressor/Compressor.cpp

```diff
@@ -0,0 +1,651 @@
+{
+	m_sampleRate = Engine::mixer()->processingSampleRate();
+
+	m_yL[0] = m_yL[1] = COMP_NOISE_FLOOR;
+
+	m_maxLookaheadVal[0] = 0;
+	m_maxLookaheadVal[1] = 0;
+
+	// 200 ms
+	m_crestTimeConst = exp(-1.f / (0.2f * m_sampleRate));
+
+	connect(&m_compressorControls.m_attackModel, SIGNAL(dataChanged()), this, SLOT(calcAttack()));
+	connect(&m_compressorControls.m_releaseModel, SIGNAL(dataChanged()), this, SLOT(calcRelease()));
+	connect(&m_compressorControls.m_holdModel, SIGNAL(dataChanged()), this, SLOT(calcHold()));
+	connect(&m_compressorControls.m_ratioModel, SIGNAL(dataChanged()), this, SLOT(calcRatio()));
+	connect(&m_compressorControls.m_rangeModel, SIGNAL(dataChanged()), this, SLOT(calcRange()));
+	connect(&m_compressorControls.m_rmsModel, SIGNAL(dataChanged()), this, SLOT(resizeRMS()));
+	connect(&m_compressorControls.m_lookaheadLengthModel, SIGNAL(dataChanged()), this, SLOT(calcLookaheadLength()));
+	connect(&m_compressorControls.m_thresholdModel, SIGNAL(dataChanged()), this, SLOT(calcThreshold()));
+	connect(&m_compressorControls.m_kneeModel, SIGNAL(dataChanged()), this, SLOT(calcKnee()));
+	connect(&m_compressorControls.m_outGainModel, SIGNAL(dataChanged()), this, SLOT(calcOutGain()));
+	connect(&m_compressorControls.m_inGainModel, SIGNAL(dataChanged()), this, SLOT(calcInGain()));
+	connect(&m_compressorControls.m_tiltModel, SIGNAL(dataChanged()), this, SLOT(calcTiltCoeffs()));
+	connect(&m_compressorControls.m_tiltFreqModel, SIGNAL(dataChanged()), this, SLOT(calcTiltCoeffs()));
+	connect(&m_compressorControls.m_limiterModel, SIGNAL(dataChanged()), this, SLOT(redrawKnee()));
+	connect(&m_compressorControls.m_mixModel, SIGNAL(dataChanged()), this, SLOT(calcMix()));
+
+	connect(&m_compressorControls.m_autoAttackModel, SIGNAL(dataChanged()), this, SLOT(calcAutoAttack()));
+	connect(&m_compressorControls.m_autoReleaseModel, SIGNAL(dataChanged()), this, SLOT(calcAutoRelease()));
+
+	connect(&m_compressorControls.m_thresholdModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+	connect(&m_compressorControls.m_ratioModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+	connect(&m_compressorControls.m_kneeModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+	connect(&m_compressorControls.m_autoMakeupModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+
+	connect(Engine::mixer(), SIGNAL(sampleRateChanged()), this, SLOT(changeSampleRate()));
+	changeSampleRate();
+}
+
+
+
+
+CompressorEffect::~CompressorEffect()
+{
+}
+
+
+float CompressorEffect::msToCoeff(float ms)
+{
+	// Convert time in milliseconds to applicable lowpass coefficient
+	return exp(m_coeffPrecalc / ms);
+}
+
+
+
+void CompressorEffect::calcAutoMakeup()
+{
+	// Formulas using the compressor's Threshold, Ratio, and Knee values to estimate a good makeup gain value
+
+	float tempGainResult;
+	if (-m_thresholdVal < -m_kneeVal)// Below knee
+	{
+		tempGainResult = 0;
+	}
+	else if (abs(1 - m_thresholdVal) < m_kneeVal)// Within knee
```

Comment:
Where does the 1 come from? isn't it just `-m_thresholdVal`?

Reply:
If I understood correctly, the separation of region is given by:
- `level < threshold - knee`: below the knee
- `threshold - knee < level < threshold + knee`: in the knee
- `threshold + knee < level`: above the knee

`level` is 0(dBFS) here.

Reply:
Oh, nice catch, I was mixing up amplitude and dBFS, which is why I used 1 instead of 0.

Reply:
Just noticed you have a similar code in `CompressorEffect::processAudioBuffer`, but you made a mistake when you assumed `currentPeakDbfs` is equivalent to 1 here.

Reply:
Also I should mention you don't need a `abs` here and Line 425.